syncer/syncer.go

```diff
@@ -288,29 +288,23 @@ func (s *Syncer) synchronize(ctx context.Context) bool {
 		s.startValidating(ctx, s.run, vQueue, vDone)
 		return nil
 	})
-	success := false
-	defer func() {
-		close(vQueue)
-		<-vDone
-		s.setStateAfterSync(ctx, success)
-		logger.With().Info(fmt.Sprintf("finished sync run #%v", s.run),
-			log.Bool("success", success),
-			log.String("sync_state", s.getSyncState().String()),
-			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
-			log.FieldNamed("latest", s.mesh.LatestLayer()),
-			log.FieldNamed("processed", s.mesh.ProcessedLayer()))
-		s.setSyncerIdle()
-	}()
 
+	success := true
+	state := s.getSyncState()
 	// using ProcessedLayer() instead of LatestLayer() so we can validate layers on a best-efforts basis.
 	// our clock starts ticking from 1 so it is safe to skip layer 0
 	// always sync to currentLayer-1 to reduce race with gossip and hare/tortoise
 	for layerID := s.mesh.ProcessedLayer().Add(1); layerID.Before(s.ticker.GetCurrentLayer()); layerID = layerID.Add(1) {
-		var layer *types.Layer
-		var err error
-		if layer, err = s.syncLayer(ctx, layerID); err != nil {
+		layer, err := s.syncLayer(ctx, layerID)
+		if err != nil {
 			logger.With().Error("failed to sync to layer", layerID, log.Err(err))
-			return false
+			success = false
+			break
+		}
+		if layer == nil {
+			// if peers don't have the layer don't revert to non-synced if node was synced
+			success = state != notSynced
```

Comment:
i don't think this make sense. the syncer always sync `currentLayer-1`. i.e. it never works on current layer. so if does get an nil layer here, that means it's not able to get data for `currentLayer-1` or older. that should set the node back to `notSynced`

Reply:
the problem is that when syncer requests currentLayer-1 that layer is not processed usually, and there is no way to know what we are actually receiving